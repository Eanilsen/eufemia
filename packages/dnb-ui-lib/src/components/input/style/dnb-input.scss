/*
* DNB Input
*
* ======================================================================== */

@import '../../../style/component/core-imports.scss';
@import '../../../style/component/font-import.scss';
@import '../../../style/core/form.scss';
@import '../../icon/style/dnb-icon.scss';

.dnb-input {
  @include componentReset();
}

@import './_input.scss';
