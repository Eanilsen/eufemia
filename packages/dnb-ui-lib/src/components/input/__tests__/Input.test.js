/**
 * Component Test
 *
 */

import React from 'react'
import {
  shallow,
  mount,
  fakeProps,
  toJson,
  loadScss
} from '../../../core/jest/jestSetup'
import Component from '../Input'
// just to make sure we re-run the test in watch mode due to changes in this file
import '../style/dnb-input.scss'

const props = {
  ...fakeProps(require.resolve('../Input'), {
    optional: true
  }),
  inputElement: null,
  disabled: false
}

describe('Input component', () => {
  // shallow compare the snapshot
  it('have to match type="text" snapshot', () => {
    const Comp = shallow(<Component {...props} type="text" value="test" />)
    expect(toJson(Comp)).toMatchSnapshot()
  })
  it('have to match type="search" snapshot', () => {
    const Comp = shallow(
      <Component {...props} type="search" value="test" />
    )
    expect(toJson(Comp)).toMatchSnapshot()
  })

  // then test the state management
  const Comp = mount(
    <Component {...props} type="text" value={null}>
      {null}
    </Component>
  )

  it('has correct state after "focus" trigger', () => {
    Comp.find('input').simulate('focus')
    // Comp.find('input').prop('onFocus')() // call onFocus handler

    expect(Comp.find('.dnb-input__shell').prop('data-input-state')).toBe(
      'focus'
    )
  })

  it('has correct state after "change" trigger', () => {
    expect(Comp.find('.dnb-input__shell').prop('data-has-content')).toBe(
      'false'
    )

    const value = 'new value'
    Comp.find('input').simulate('change', { target: { value } })

    expect(Comp.find('.dnb-input__shell').prop('data-has-content')).toBe(
      'true'
    )

    expect(Comp.state().value).toBe(value)
  })

  // make sure getDerivedStateFromProps works
  it('has correct state after changeing "value" prop (set by getDerivedStateFromProps)', () => {
    const value = 'new prop value'
    Comp.setProps({
      value
    })
    expect(Comp.state().value).toBe(value)

    // get dom node
    // console.log('domNode', Comp.find('input').getDOMNode().value)
  })

  it(`has to to have a prop value like value`, () => {
    const value = 'new value'
    Comp.setProps({
      value
    })
    expect(Comp.find('input').props().value).toBe(value)
  })

  it('has a submit button on prop type="search"', () => {
    const Comp = mount(
      <Component {...props} type="search" value={null}>
        {null}
      </Component>
    )

    const Button = Comp.find('Submit').find('button')
    expect(Button.exists()).toBe(true)

    Button.simulate('focus')
    expect(
      Comp.find('Submit')
        .find('.dnb-input__search-submit')
        .prop('data-input-state')
    ).toBe('focus')
  })
})

describe('Input scss', () => {
  it('have to match snapshot', () => {
    const scss = loadScss(require.resolve('../style/dnb-input.scss'))
    expect(scss).toMatchSnapshot()
  })
})
