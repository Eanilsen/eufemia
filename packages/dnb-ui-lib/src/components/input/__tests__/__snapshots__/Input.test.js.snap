// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input component have to match type="search" snapshot 1`] = `
<span
  className="dnb-input dnb-input--search dnb-input--has-extra-information dnb-input--size class className"
>
  <span
    className="dnb-input__shell"
    data-has-content="true"
    data-input-state="virgin"
  >
    <input
      aria-describedby="id-description"
      autoComplete="autocomplete"
      className="dnb-input__input font_class dnb-input__input--search dnb-input__align--align"
      id="id"
      name="id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      type="search"
      value="test"
    />
    <Submit
      align="align"
      attributes="attributes"
      autocomplete="autocomplete"
      class="class"
      className="className"
      custom_element={Object {}}
      custom_method={[Function]}
      description="description"
      disabled={false}
      extra_information="extra_information"
      font_class="font_class"
      icon_size={16}
      id="id"
      inputElement={null}
      onSubmit={null}
      onSubmitBlur={null}
      onSubmitFocus={null}
      on_blur={[Function]}
      on_change={[Function]}
      on_focus={[Function]}
      on_submit={[Function]}
      on_submit_blur={[Function]}
      on_submit_focus={[Function]}
      placeholder="placeholder"
      search_button_title="search_button_title"
      size="size"
      title="search_button_title"
      type="search"
      value="test"
    >
      children
    </Submit>
    <span
      className="dnb-input__placeholder font_class dnb-input__align--align"
    >
      placeholder
    </span>
  </span>
  <span
    className="dnb-input__description"
    id="id-description"
  >
    description
  </span>
  <span
    className="dnb-input__extra-information"
  >
    extra_information
  </span>
</span>
`;

exports[`Input component have to match type="text" snapshot 1`] = `
<span
  className="dnb-input dnb-input--text dnb-input--has-extra-information dnb-input--size class className"
>
  <span
    className="dnb-input__shell"
    data-has-content="true"
    data-input-state="virgin"
  >
    <input
      aria-describedby="id-description"
      autoComplete="autocomplete"
      className="dnb-input__input font_class dnb-input__align--align"
      id="id"
      name="id"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      type="text"
      value="test"
    />
    <span
      className="dnb-input__placeholder font_class dnb-input__align--align"
    >
      placeholder
    </span>
  </span>
  <span
    className="dnb-input__description"
    id="id-description"
  >
    description
  </span>
  <span
    className="dnb-input__extra-information"
  >
    extra_information
  </span>
</span>
`;

exports[`Input scss have to match snapshot 1`] = `
"/*
* DNB Input
*
* ======================================================================== */
/*
 * Utilities
 * ========================================================================== */
/*
 * Typography
 * ========================================================================== */
:root {
  --font-family-std: 'Fedra Sans Std', sans-serif;
  --font-family-demi: 'Fedra Sans Std Demi', sans-serif; }

.typo-std,
.typo-book {
  font-family: var(--font-family-std);
  font-weight: normal;
  font-style: normal; }

.typo-medium {
  font-family: var(--font-family-std);
  font-weight: 600;
  font-style: normal; }

.typo-demi {
  font-family: var(--font-family-demi);
  font-weight: normal;
  font-style: normal; }

.typo-number--old-style {
  font-variant-numeric: oldstyle-nums;
  -moz-font-feature-settings: \\"onum\\";
  -webkit-font-feature-settings: \\"onum\\";
  font-feature-settings: \\"onum\\"; }

.typo-number--lining {
  font-variant-numeric: lining-nums;
  -moz-font-feature-settings: \\"lnum\\";
  -webkit-font-feature-settings: \\"lnum\\";
  font-feature-settings: \\"lnum\\"; }

.typo-number--tabular {
  font-variant-numeric: tabular-nums;
  -moz-font-feature-settings: \\"tnum\\";
  -webkit-font-feature-settings: \\"tnum\\";
  font-feature-settings: \\"tnum\\"; }

.typo-number--proportional {
  font-variant-numeric: proportional-nums;
  -moz-font-feature-settings: \\"pnum\\";
  -webkit-font-feature-settings: \\"pnum\\";
  font-feature-settings: \\"pnum\\"; }

@font-face {
  font-family: 'Fedra Sans Std';
  src: url(\\"../../../../assets/fonts//FedraSansStd-Book.eot\\");
  src: url(\\"../../../../assets/fonts//FedraSansStd-Book.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts//FedraSansStd-Book.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts//FedraSansStd-Book.ttf\\") format(\\"truetype\\");
  font-weight: normal;
  font-style: normal; }

@font-face {
  font-family: 'Fedra Sans Std Demi';
  src: url(\\"../../../../assets/fonts//FedraSansStd-Demi.eot\\");
  src: url(\\"../../../../assets/fonts//FedraSansStd-Demi.eot?#iefix\\") format(\\"embedded-opentype\\"), url(\\"../../../../assets/fonts//FedraSansStd-Demi.woff\\") format(\\"woff\\"), url(\\"../../../../assets/fonts//FedraSansStd-Demi.ttf\\") format(\\"truetype\\");
  font-weight: normal;
  font-style: normal; }

/*
* Scopes
* ======================================================================== */
/*
* DNB Button
*
* ======================================================================== */
/*
* DNB icon
*
* ======================================================================== */
/*
* Icon component
*
* ======================================================================== */
.dnb-icon {
  display: inline;
  vertical-align: baseline;
  font-size: inherit;
  width: 1em;
  height: 1em; }
  .dnb-icon svg {
    fill: currentColor; }
  .dnb-icon:not(.has-custom-size) svg {
    width: inherit;
    height: inherit; }
  .dnb-icon--small {
    width: 0.5em;
    height: 0.5em; }
  .dnb-icon--default {
    width: 1em;
    height: 1em; }
  .dnb-icon--medium {
    width: 1.5em;
    height: 1.5em; }
  .dnb-icon--large {
    width: 2em;
    height: 2em; }

.dnb-button {
  font-family: var(--font-family-std, sans-serif);
  font-size: 1rem;
  font-weight: normal;
  font-style: normal;
  line-height: 1.5rem;
  color: var(--color-black-80, #333); }
  .dnb-button *,
  .dnb-button ::before,
  .dnb-button ::after {
    background-repeat: no-repeat;
    box-sizing: border-box; }
  .dnb-button ::before,
  .dnb-button ::after {
    text-decoration: inherit;
    vertical-align: inherit; }

/*
* Button component
*
* ======================================================================== */
:root {
  --button-font-family: var(--font-family-std);
  --button-font-size: 1rem;
  --button-font-weight: normal;
  --button-font-style: normal;
  --button-width: 2.5rem;
  --button-height: 2.5rem;
  --button-width--medium: 2rem;
  --button-height--medium: 2rem;
  --button-width--large: 3rem;
  --button-height--large: 3rem;
  --button-icon-size: 1rem;
  --button-border-width: 1px;
  --button-border-width--hover: 3px;
  --button-border-radius: calc(var(--button-height) / 2);
  --button-border-radius--medium: calc(var(--button-height--medium) / 2);
  --button-border-radius--large: calc(var(--button-height--large) / 2); }

.dnb-button {
  display: inline-flex;
  vertical-align: middle;
  align-content: center;
  align-items: center;
  justify-content: center;
  white-space: nowrap;
  cursor: pointer;
  width: var(--button-width);
  height: var(--button-height);
  padding: 0;
  border: var(--button-border-width) solid transparent;
  border-radius: var(--button-border-radius);
  font-family: var(--button-font-family);
  font-size: var(--button-font-size);
  font-weight: var(--button-font-weight);
  font-style: var(--button-font-style);
  text-decoration: none;
  /* Firefox includes a hidden border which messes up button dimensions */ }
  .dnb-button__icon {
    line-height: 0; }
    .dnb-button__icon:not(.dnb-icon) svg:not([width]):not([height]) {
      width: var(--button-icon-size);
      height: var(--button-icon-size); }
  .dnb-button__text {
    color: inherit;
    background-color: inherit;
    transform: translateY(var(--button-border-width)); }
  .dnb-button--size-medium {
    width: var(--button-width--medium);
    height: var(--button-height--medium);
    border-radius: var(--button-border-radius--medium); }
  .dnb-button--has-text.dnb-button--size-medium {
    padding-left: 1em;
    padding-right: 1em; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-medium {
    padding-left: 0.5em; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-medium {
    padding-right: 0.5em; }
  .dnb-button--size-large {
    width: var(--button-width--large);
    height: var(--button-height--large);
    border-radius: var(--button-border-radius--large); }
  .dnb-button--has-text.dnb-button--size-large {
    padding-left: 2em;
    padding-right: 2em; }
  .dnb-button--has-text.dnb-button--icon-position-left.dnb-button--size-large {
    padding-left: 1em; }
  .dnb-button--has-text.dnb-button--icon-position-right.dnb-button--size-large {
    padding-right: 1em; }
  .dnb-button--has-text {
    padding-left: 1.5em;
    padding-right: 1.5em;
    width: auto; }
  .dnb-button--has-text .dnb-button__icon {
    margin: 0 calc(var(--button-icon-size) / 2); }
  .dnb-button--has-text.dnb-button--icon-position-left {
    padding-left: 0.5em; }
  .dnb-button--has-text.dnb-button--icon-position-right {
    padding-right: 0.5em; }
  .dnb-button--has-text.dnb-button--has-icon .dnb-button__icon {
    order: 2; }
  .dnb-button--has-text.dnb-button--has-icon .dnb-button__text {
    order: 1; }
  .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__icon {
    order: 1; }
  .dnb-button--has-text.dnb-button--has-icon.dnb-button--icon-position-left .dnb-button__text {
    order: 2; }
  .dnb-button--reset {
    margin: 0;
    padding: 0;
    width: auto;
    height: auto;
    overflow: visible;
    border: none;
    border-radius: 0;
    background-color: transparent;
    appearance: none;
    box-shadow: none;
    color: inherit;
    font: inherit;
    text-align: inherit;
    line-height: 0; }
    .no-touch .dnb-button--reset:hover[disabled] {
      cursor: not-allowed; }
    .no-touch .dnb-button--reset:hover:not([disabled]) {
      opacity: 0.6;
      box-shadow: none;
      border: none; }
    .dnb-button--reset:not([disabled]):focus, .dnb-button--reset:not([disabled]):active {
      outline: none;
      border-color: transparent; }
      html[data-whatinput='mouse'] .dnb-button--reset:not([disabled]):focus:not(.dnb-no-mouse-focus), html[data-whatinput='mouse'] .dnb-button--reset:not([disabled]):active:not(.dnb-no-mouse-focus) {
        transition: box-shadow 0.1s ease-out 0s;
        box-shadow: 0 0 0 2px var(--color-emerald-green); }
      html[data-whatinput='keyboard'] .dnb-button--reset:not([disabled]):focus, html[data-whatinput='keyboard'] .dnb-button--reset:not([disabled]):active {
        box-shadow: 0 0 0 2px var(--color-signal-orange); }
    html[data-whatinput='mouse']
.dnb-button--reset:not([disabled]):focus:not(.dnb-no-mouse-focus),
    html[data-whatinput='mouse']
.dnb-button--reset:not([disabled]):active:not(.dnb-no-mouse-focus) {
      box-shadow: none;
      color: inherit;
      border: none; }
  .dnb-button[type='button'] {
    appearance: none; }
  a .dnb-button {
    border-bottom: none; }
  button.dnb-button::-moz-focus-inner .dnb-button {
    border: none; }

.dnb-input {
  font-family: var(--font-family-std, sans-serif);
  font-size: 1rem;
  font-weight: normal;
  font-style: normal;
  line-height: 1.5rem;
  color: var(--color-black-80, #333); }
  .dnb-input *,
  .dnb-input ::before,
  .dnb-input ::after {
    background-repeat: no-repeat;
    box-sizing: border-box; }
  .dnb-input ::before,
  .dnb-input ::after {
    text-decoration: inherit;
    vertical-align: inherit; }

/*
* Input component
*
* ======================================================================== */
/*
* Forms and inputs
* ======================================================================== */
.dnb-input {
  display: flex;
  align-items: center;
  flex-flow: row wrap;
  width: 100%; }
  .dnb-input--label-hidden label {
    border: 0;
    clip: rect(0 0 0 0);
    height: 1px;
    margin: -1px;
    overflow: hidden;
    padding: 0;
    position: absolute;
    width: 1px; }
  .dnb-input__shell {
    position: relative;
    flex: 1;
    display: flex;
    width: 100%;
    height: 100%;
    margin: 0;
    align-items: center;
    background-color: #f5fafa;
    border-radius: 2rem;
    border: 1px solid #bedbdb;
    transition: background-color 0.5s ease-out 0s; }
    .no-touch .dnb-input__shell input:not([disabled]):hover[disabled] {
      cursor: not-allowed; }
    .no-touch .dnb-input__shell input:not([disabled]):hover:not([disabled]) {
      background-color: rgba(255, 255, 255, 0.7); }
  .dnb-input__description, .dnb-input__placeholder, .dnb-input__input {
    font-size: 1rem;
    line-height: 1.25; }
  .dnb-input__placeholder, .dnb-input__input {
    width: 100%;
    height: 2.2rem;
    padding: 1.2rem 0 1rem 1rem;
    font-family: inherit;
    font-weight: inherit;
    outline: none;
    background-color: transparent;
    color: inherit;
    border: none;
    border-radius: 2rem; }
  .dnb-input__input[disabled] {
    opacity: 0.5;
    user-select: none; }
  .dnb-input__placeholder {
    pointer-events: none;
    position: absolute;
    left: 0;
    top: 2px;
    padding: 0 1rem;
    height: 100%;
    overflow: hidden;
    white-space: nowrap;
    opacity: 1;
    color: rgba(var(--color-black-80), 0.8);
    line-height: 2.2rem;
    text-overflow: ellipsis;
    border-radius: 0;
    transition: opacity 0.1s linear 0s; }
  .dnb-input__description {
    margin: 0 0 0 0.5rem; }
  .dnb-input__extra-information {
    flex: 1 100%;
    margin: 0.5rem 0 0 1rem;
    font-size: 0.75rem;
    line-height: 1rem; }
  .dnb-input--large .dnb-input__description,
  .dnb-input--large .dnb-input__input,
  .dnb-input--large .dnb-input__placeholder {
    font-size: 2rem; }
  .dnb-input--large .dnb-input__placeholder {
    line-height: 3rem; }
  .dnb-input--large .dnb-input__input {
    height: 3rem;
    padding: 1.5rem 0 1.5rem 1.5rem;
    border-radius: 3rem; }
  .dnb-input__align--right {
    padding-left: 0 !important;
    padding-right: 1rem !important;
    text-align: right; }
  .dnb-input__input--search {
    border-radius: 2rem 0 0 2rem;
    margin: 0 1px 0 0; }
  .dnb-input__search-submit {
    position: relative;
    top: -1px;
    right: -1px;
    /* stylelint-disable */
    flex: 0 0 2.2rem;
    flex: 0 0 calc(4px + 2.2rem);
    height: 2.2rem;
    height: calc(4px + 2.2rem);
    /* stylelint-enable */
    margin: 0 0 -2px; }
    .dnb-input__search-submit__button {
      width: 100%;
      height: 100%; }
  .dnb-input__shell[data-input-state='focus'] {
    border-color: #a0d6c9;
    background-color: white;
    box-shadow: 0 0 0 2px var(--color-sea-green);
    outline: none;
    border-color: transparent; }
    html[data-whatinput='mouse'] .dnb-input__shell[data-input-state='focus']:not(.dnb-no-mouse-focus) {
      transition: box-shadow 0.1s ease-out 0s;
      box-shadow: 0 0 0 2px var(--color-emerald-green); }
    html[data-whatinput='keyboard'] .dnb-input__shell[data-input-state='focus'] {
      box-shadow: 0 0 0 2px var(--color-signal-orange); }
  .dnb-input__shell[data-has-content='true'] .dnb-input__placeholder,
  .dnb-input__shell[data-input-state='focus'][data-has-content='true'] .dnb-input__placeholder {
    display: none; }
  .dnb-input__shell[data-input-state='focus'][data-has-content='false'] .dnb-input__placeholder {
    opacity: 0.35; }

.dnb-input__shell[data-has-content='true'] {
  background-color: rgba(255, 255, 255, 0.7); }

/*
* Range input
* ======================================================================== */
.dnb-range-slider__css-vars__bg-color-1 {
  background-color: rgba(var(--color-sea-green), 0.5); }

.dnb-range-slider__css-vars__bg-color-2 {
  background-color: var(--color-summer-green); }

input[type='range'] {
  -webkit-appearance: none;
  width: 100%;
  border-radius: 0.25rem; }

input[type='range']::-webkit-slider-runnable-track {
  position: relative;
  z-index: 1;
  cursor: pointer;
  width: 100%;
  height: 0.25rem;
  box-shadow: none; }

input[type='range']::-webkit-slider-thumb {
  position: relative;
  top: 0.125rem;
  cursor: pointer;
  height: 1.5rem;
  width: 1.5rem;
  border-radius: 1.5rem;
  background: white;
  border: 1px solid #98c6c6;
  transform: translate(0, -50%);
  -webkit-appearance: none;
  box-shadow: 1px 1px 2px rgba(0, 135, 137, 0.25), 0 0 1px rgba(0, 160, 163, 0.25); }

input[type='range']:focus {
  outline: none; }
"
`;
